pkg/controller/mpi_job_controller.go

```diff
@@ -961,8 +961,13 @@ func (c *MPIJobController) getOrCreateWorker(mpiJob *kubeflow.MPIJob) ([]*corev1
 		// If an error occurs during Get/Create, we'll requeue the item so we
 		// can attempt processing again later. This could have been caused by a
 		// temporary network failure, or any other transient reason.
+		// But, if err is about pod spec invalid, retrying would be
+		// futile, the status of job should turn to failed.
 		if err != nil {
 			c.recorder.Eventf(mpiJob, corev1.EventTypeWarning, mpiJobFailedReason, "worker pod created failed: %v", err)
```

Comment:
This is only one of the cases where there could be an invalid Pod template.

It might be better to return this error and handle more generically in syncHandler, so we can handle the launcher pod, the worker pods and any other validation errors:
https://github.com/kubeflow/mpi-operator/blob/6748c47ba369e570ec0eea8c4e5c44b08d48bd3e/pkg/controller/mpi_job_controller.go#L562

Reply:
Agree.

Reply:
I have examined how Pod Spec validation is performed in the Kubernetes project. The relevant code can be found in the "k8s.io/kubernetes/pkg/apis/core/validation" package. 
However, it seems that this package is not usable outside of the Kubernetes project

Reply:
I didn't mean that you should use the validation code form kubernetes.

I just mean that there are multiple cases in which we can't retry, and this PR is only covering one of them.